File: projects/chess-coach/apps/harness/src/main/kotlin/harness/LlmClient.kt
package harness

import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.plugins.HttpTimeout
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import io.ktor.utils.io.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class ChatMessage(val role: String, val content: String)

@Serializable
data class ChatRequest(
    val model: String,
    val messages: List<ChatMessage>,
    val stream: Boolean = false,
    val temperature: Double = 0.7,
    val format: String? = null,
    @SerialName("max_tokens") val maxTokens: Int? = null
)

@Serializable
data class ChatChoice(val message: ChatMessage)

@Serializable
data class ChatResponse(val choices: List<ChatChoice>)

@Serializable
data class ChatChunkChoice(val delta: ChatMessage)

@Serializable
data class ChatChunkResponse(val choices: List<ChatChunkChoice>)

class LlmClient {
    // Configuration via Environment Variables
    private val baseUrl: String = System.getenv("LLM_BASE_URL") 
        ?: System.getenv("OLLAMA_BASE_URL") 
        ?: "http://localhost:11434/v1"
        
    private val apiKey: String? = System.getenv("LLM_API_KEY")
    
    val generalModel: String = System.getenv("LLM_GENERAL_MODEL") ?: System.getenv("LLM_MODEL") ?: "qwen2.5:7b"
    val commentaryModel: String = System.getenv("LLM_COMMENTARY_MODEL") ?: generalModel

    private val client = HttpClient(CIO) {
        install(ContentNegotiation) {
            json(Json {
                ignoreUnknownKeys = true
                prettyPrint = true
            })
        }
        install(HttpTimeout) {
            requestTimeoutMillis = 30_000
            connectTimeoutMillis = 30_000
            socketTimeoutMillis = 30_000
        }
    }

    suspend fun prompt(
        systemPrompt: String, 
        userPrompt: String, 
        model: String = commentaryModel,
        temperature: Double = 0.7,
        format: String? = null,
        maxTokens: Int? = null
    ): String {
        val requestPayload = ChatRequest(
            model = model,
            messages = listOf(
                ChatMessage(role = "system", content = systemPrompt),
                ChatMessage(role = "user", content = userPrompt)
            ),
            temperature = temperature,
            format = format,
            maxTokens = maxTokens
        )

        val response: ChatResponse = client.post("$baseUrl/chat/completions") {
            contentType(ContentType.Application.Json)
            
            // Inject API Key if provided (Required for OpenAI, Groq, etc.)
            if (!apiKey.isNullOrBlank()) {
                bearerAuth(apiKey)
            }
            
            setBody(requestPayload)
        }.body()

        return response.choices.firstOrNull()?.message?.content ?: ""
    }

    suspend fun promptStream(
        systemPrompt: String, 
        userPrompt: String, 
        model: String = commentaryModel,
        temperature: Double = 0.7,
        maxTokens: Int? = null
    ): Flow<String> = flow {
        val requestPayload = ChatRequest(
            model = model,
            messages = listOf(
                ChatMessage(role = "system", content = systemPrompt),
                ChatMessage(role = "user", content = userPrompt)
            ),
            stream = true,
            temperature = temperature,
            maxTokens = maxTokens
        )

        val jsonParser = Json { ignoreUnknownKeys = true }

        client.preparePost("$baseUrl/chat/completions") {
            contentType(ContentType.Application.Json)
            if (!apiKey.isNullOrBlank()) {
                bearerAuth(apiKey)
            }
            setBody(requestPayload)
        }.execute { response ->
            val channel = response.bodyAsChannel()
            while (!channel.isClosedForRead) {
                val line = channel.readUTF8Line() ?: break
                if (line.startsWith("data: ") && line != "data: [DONE]") {
                    val jsonStr = line.removePrefix("data: ")
                    try {
                        val chunk = jsonParser.decodeFromString<ChatChunkResponse>(jsonStr)
                        val content = chunk.choices.firstOrNull()?.delta?.content ?: ""
                        if (content.isNotEmpty()) {
                            emit(content)
                        }
                    } catch (e: Exception) {
                        // Ignore parse errors for partial/malformed chunks
                    }
                }
            }
        }
    }
    
    fun close() {
        client.close()
    }
}

File: projects/chess-coach/apps/server/src/LlmClient.ts
/**
 * Commentary LLM transport. Mirrors the Kotlin LlmClient
 * (_spec/api/behavior.md §6): OpenAI-compatible POST {base}/chat/completions,
 * unary + SSE streaming, optional bearer auth.
 */
export interface LlmClient {
  readonly commentaryModel: string;
  prompt(
    systemPrompt: string,
    userPrompt: string,
    model?: string,
    temperature?: number,
    maxTokens?: number,
  ): Promise<string>;
  promptStream(
    systemPrompt: string,
    userPrompt: string,
    model?: string,
    temperature?: number,
    maxTokens?: number,
  ): AsyncGenerator<string, void, unknown>;
}

type ChatCompletion = { choices?: Array<{ message?: { content?: string } }> };
type ChatChunk = { choices?: Array<{ delta?: { content?: string } }> };

export class OllamaLlmClient implements LlmClient {
  private readonly baseUrl: string;
  private readonly apiKey?: string;
  readonly generalModel: string;
  readonly commentaryModel: string;

  constructor() {
    this.baseUrl =
      process.env.LLM_BASE_URL ?? process.env.OLLAMA_BASE_URL ?? "http://localhost:11434/v1";
    this.apiKey = process.env.LLM_API_KEY || undefined;
    this.generalModel = process.env.LLM_GENERAL_MODEL ?? process.env.LLM_MODEL ?? "qwen2.5:7b";
    this.commentaryModel = process.env.LLM_COMMENTARY_MODEL ?? this.generalModel;
  }

  private headers(): Record<string, string> {
    const h: Record<string, string> = { "Content-Type": "application/json" };
    if (this.apiKey) h["Authorization"] = `Bearer ${this.apiKey}`;
    return h;
  }

  async prompt(
    systemPrompt: string,
    userPrompt: string,
    model: string = this.commentaryModel,
    temperature = 0.7,
    maxTokens?: number,
  ): Promise<string> {
    const res = await fetch(`${this.baseUrl}/chat/completions`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify({
        model,
        messages: [
          { role: "system", content: systemPrompt },
          { role: "user", content: userPrompt },
        ],
        temperature,
        max_tokens: maxTokens,
      }),
    });
    if (!res.ok) throw new Error(`LLM request failed: ${res.status}`);
    const data = (await res.json()) as ChatCompletion;
    return data.choices?.[0]?.message?.content ?? "";
  }

  async *promptStream(
    systemPrompt: string,
    userPrompt: string,
    model: string = this.commentaryModel,
    temperature = 0.7,
    maxTokens?: number,
  ): AsyncGenerator<string, void, unknown> {
    const res = await fetch(`${this.baseUrl}/chat/completions`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify({
        model,
        messages: [
          { role: "system", content: systemPrompt },
          { role: "user", content: userPrompt },
        ],
        stream: true,
        temperature,
        max_tokens: maxTokens,
      }),
    });
    if (!res.ok || !res.body) throw new Error(`LLM stream failed: ${res.status}`);

    const reader = res.body.getReader();
    const decoder = new TextDecoder();
    let buf = "";
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      buf += decoder.decode(value, { stream: true });
      let idx: number;
      while ((idx = buf.indexOf("\n")) >= 0) {
        const line = buf.slice(0, idx).trim();
        buf = buf.slice(idx + 1);
        if (line.startsWith("data: ") && line !== "data: [DONE]") {
          try {
            const chunk = JSON.parse(line.slice(6)) as ChatChunk;
            const content = chunk.choices?.[0]?.delta?.content ?? "";
            if (content) yield content;
          } catch {
            // ignore partial/malformed chunk
          }
        }
      }
    }
  }
}
